src/radical/pilot/agent/agent_0.py

```diff
@@ -180,110 +210,68 @@ def initialize_parent(self):
-        self._prof.prof(event='hostname', uid=self._pid, msg=ru.get_hostname())
+        # all work is done in the registered callbacks
+        time.sleep(1)
 
 
     # --------------------------------------------------------------------------
     #
-    def finalize_parent(self):
+    def finalize(self):
 
         # tear things down in reverse order
-        self._prof.flush()
-        self._log.info('publish "terminate" cmd')
-        self.publish(rpc.CONTROL_PUBSUB, {'cmd' : 'terminate',
-                                          'arg' : None})
-
-        self.unregister_timed_cb(self._check_units_cb)
-        self.unregister_output(rps.AGENT_STAGING_INPUT_PENDING)
-        self.unregister_timed_cb(self._agent_command_cb)
+        self._hb.stop()
+        self._cmgr.close()
 
         if self._lrms:
-            self._log.debug('stop    lrms %s', self._lrms)
             self._lrms.stop()
-            self._log.debug('stopped lrms %s', self._lrms)
 
         if   self._final_cause == 'timeout'  : state = rps.DONE
         elif self._final_cause == 'cancel'   : state = rps.CANCELED
         elif self._final_cause == 'sys.exit' : state = rps.CANCELED
         else                                 : state = rps.FAILED
 
-        self._log.debug('final state: %s (%s)', state, self._final_cause)
-
         # we don't rely on the existence / viability of the update worker at
         # that point.
         self._log.debug('update db state: %s: %s', state, self._final_cause)
         self._update_db(state, self._final_cause)
 
-
-    # --------------------------------------------------------------------------
-    #
-    def wait_final(self):
-
-        while self._final_cause is None:
-          # self._log.info('no final cause -> alive')
-            time.sleep(1)
-
-        self._log.debug('final: %s', self._final_cause)
-
-      # if self._session:
-      #     self._log.debug('close  session %s', self._session.uid)
-      #     self._session.close()
-      #     self._log.debug('closed session %s', self._session.uid)
-
-
-    # --------------------------------------------------------------------------
-    #
-    def _update_db(self, state, msg=None):
-
         # NOTE: we do not push the final pilot state, as that is done by the
-        #       bootstrapper *after* this poilot *actually* finished.
-
-        self._log.info('pilot state: %s', state)
-        self._log.info('rusage: %s', rpu.get_rusage())
-        self._log.info(msg)
+        #       bootstrapper *after* this pilot *actually* finished.
 
-        if state == rps.FAILED:
-            self._log.info(ru.get_trace())
+        self._log.info('pilot state: %s [%s]', state, self._final_cause)
 
-        out = None
-        err = None
-        log = None
+        out, out, err = None, None, None
 
-        try   : out = open('./agent_0.out', 'r').read(1024)
+        try   : out = open('./agent.0.out', 'r').read(1024)
         except: pass
```

Comment:
Should we raise/log a warning?

Reply:
No, that is actually not an error - the agent may well have not printed any output, we just check.  Apparently python people like to do it like this, I personally would prefer:
```
if os.file.isfile(fname) and os.file.size(fname): out = open(fname, 'r').read(1024)
```
oh well...